### PR TITLE
Derivative of matrix expression is now implicit

### PR DESCRIPTION
**Pull request by Upabjojr**
Got rid of all parsers of indexed expressions for the calculation of the matrix derivative.

#### Release Notes

* matrices
  * Derivatives of matrix expressions do not rely on indexed expressions anymore. The algorithm now calculates the result implicitly.

### CODE REVIEW

**Comment by Upabjojr** [comment]
@asmeurer the matrix derivation algorithm got suddenly much simpler.

**Comment by Upabjojr** [comment]
Now we need many more tests.

File: sympy/matrices/expressions/matexpr.py [hunk 1]
```diff
@@ -821,6 +778,133 @@ def __nonzero__(self):
...
+    second = lines[0].second
+
+    one_final = (first.shape[1] == 1) and (second.shape[1] == 1)
+
+    if lines[0].trace or one_final:
+        return reduce(lambda x,y: x+y, [lr.first * lr.second.T + lr.first_T * lr.second_T.T for lr in lines])
+
+    shape = first.shape + second.shape
+    rank = sum([i != 1 for i in shape])
+    if rank > 2:
```
Comment (asmeurer) [inline sympy/matrices/expressions/matexpr.py]: It would probably be useful to be able to return this in some form for rank > 2.

**Comment by asmeurer** [comment]
I fixed what I am assuming was a typo in the release notes entry.

File: sympy/matrices/expressions/matexpr.py [hunk 2]
```diff
@@ -821,6 +778,133 @@ def __nonzero__(self):
...
+            return False
+        return (self.first == other.first) and (self.second == other.second) and (self.first_T == other.first_T) and (self.second_T == other.second_T)
+
+
+def _matrix_derivative(expr, x):
+    from .trace import Trace
+    from .transpose import Transpose
+    from sympy import (Identity, ZeroMatrix)
+
+    if isinstance(expr, MatrixSymbol):
```
Comment (asmeurer) [inline sympy/matrices/expressions/matexpr.py]: Can we move this code into the respective _eval_derivative methods of each class?

Comment (asmeurer) [reply sympy/matrices/expressions/matexpr.py]: I guess each class should have a `_eval_derivative_lines` or something like that and the base class MatrixExpr would implement `_eval_derivative` would be your `matrix_derivative` function.

**Comment by asmeurer** [comment]
What should be done with MatrixExpr.from_index_summation? It apparently has many wrong result bugs. Can it easily be replaced with the code you wrote in sympy.codegen?

File: sympy/matrices/expressions/matexpr.py [hunk 3]
```diff
@@ -821,6 +778,133 @@ def __nonzero__(self):
...
+        return _matrix_derivative(newexpr, x)
+    if isinstance(expr, Transpose):
+        lines = _matrix_derivative(expr.args[0], x)
+        return [i.transpose() for i in lines]
+    if isinstance(expr, Trace):
+        r = _matrix_derivative(expr.args[0], x)
+        for lr in r:
+            lr.trace = True
+        return r
+    raise NotImplementedError
```
Comment (asmeurer) [inline sympy/matrices/expressions/matexpr.py]: This should result in an unevaluated derivative.

**Comment by asmeurer** [comment]
This is interesting. Is this the algorithm from the matrix calculus paper (http://www.matrixcalculus.org/)? Is their code online somewhere? 

The algorithm from that paper seems not too different from what you did originally, except it uses Einstein notation instead of explicit summations. Perhaps also the explicit use of contravariant and covariant indices helps as well?

File: sympy/matrices/expressions/matexpr.py (hunk 2 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/matexpr.py]: > Can we move this code into the respective _eval_derivative methods of each class?

Theoretically yes, I have developed the code in one block for simplicity. I would first expand the tests a lot to make sure that everything works.

I need very long expressions with intertwines expression trees to check whether the visit of the expression tree is correct.

Comment (asmeurer) [reply sympy/matrices/expressions/matexpr.py]: You can generate random expressions, but is there any way to check that the result is correct? I guess you can check it against the matrix calculus online calculator.

File: sympy/matrices/expressions/matexpr.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/matexpr.py]: Yes, although most higher rank tensors are hard to represent in terms of operators on matrices (unless you go the index explicit way).

File: sympy/matrices/expressions/matexpr.py (hunk 2 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/matexpr.py]: Yes, there is that website. And we can still check by going the explicit-index way with Kronecker deltas.

**Comment by Upabjojr** [comment]
> Perhaps also the explicit use of contravariant and covariant indices helps as well?

Contravariant and covariant here are really an abuse of terminology. If I understand correctly, the authors of MatrixCalculus.org use that convention to track the transpositions.

This code is not that different from the index notation, it simply predicts the matrix lines and follows them without creating indices.

**Comment by asmeurer** [comment]
I see. So creating the expression with the indices and KroneckerDeltas to check the answer should work then, because it's not just doing the same thing as the algorithm itself.

**Comment by Upabjojr** [comment]
Yes, matrix elements with indices are exactly the same concept as matrix expressions. The only difference is whether or not you want to see the indices.

**Comment by Upabjojr** [comment]
We also need to finish copying the cookbook examples.

**Comment by Upabjojr** [comment]
Added some more tests from the cookbooks. Two tests are not passing :(

File: sympy/matrices/expressions/matexpr.py [hunk 4]
```diff
@@ -821,6 +778,132 @@ def __nonzero__(self):
 def matrix_symbols(expr):
     return [sym for sym in expr.free_symbols if sym.is_Matrix]
 
+
+class _LeftRightArgs(object):
```
Comment (asmeurer) [inline sympy/matrices/expressions/matexpr.py]: If there is any reference for the algorithm can you add it somewhere as a comment or docstring.

File: sympy/matrices/expressions/tests/test_derivatives.py [hunk 5]
```diff
@@ -189,3 +200,43 @@ def test_matrix_derivatives_of_traces():
...
+
+    # Cookbook example 127:
+    expr = Trace((A + X.T*C*X).inv()*(X.T*B*X))
+    #assert expr.diff(X) == -2*C*X*(A + X.T*C*X).inv()*X.T*B*X*(A + X.T*C*X).inv() + 2*B*X*(A + X.T*C*X).inv()
+
+
+def test_derivatives_of_complicated_matrix_expr():
+    expr = a.T*(A*X*(X.T*B + X*A) + B.T*X.T*(a*b.T*(X*D*X.T + X*(X.T*B + A*X)*D*B - X.T*C.T*A)*B + B*(X*D.T + B*A*X*A.T - 3*X*D))*B + 42*X*B*X.T*A.T*(X + X.T))*b
+    result = (B*(B*A*X*A.T - 3*X*D + X*D.T) + a*b.T*(X*(A*X + X.T*B)*D*B + X*D*X.T - X.T*C.T*A)*B)*B*b*a.T*B.T + B**2*b*a.T*B.T*X.T*a*b.T*X*D + 42*A*X*B.T*X.T*a*b.T + B*D*B**3*b*a.T*B.T*X.T*a*b.T*X + B*b*a.T*A*X + 42*a*b.T*(X + X.T)*A*X*B.T + b*a.T*X*B*a*b.T*B.T**2*X*D.T + b*a.T*X*B*a*b.T*B.T**3*D.T*(B.T*X + X.T*A.T) + 42*b*a.T*X*B*X.T*A.T + 42*A.T*(X + X.T)*b*a.T*X*B + A.T*B.T**2*X*B*a*b.T*B.T*A + A.T*a*b.T*(A.T*X.T + B.T*X) + A.T*X.T*b*a.T*X*B*a*b.T*B.T**3*D.T + B.T*X*B*a*b.T*B.T*D - 3*B.T*X*B*a*b.T*B.T*D.T - C.T*A*B**2*b*a.T*B.T*X.T*a*b.T + X.T*A.T*a*b.T*A.T
+    assert expr.diff(X) == result
```
Comment (asmeurer) [inline sympy/matrices/expressions/tests/test_derivatives.py]: Can you add a test against the indexed expressions, to verify that the result is actually correct.

**Comment by asmeurer** [comment]
This is looking nice. I'd like to start a release branch before the end of the year, if possible. How long do you think it will take to finish this and get it merged? I would like to get this in the next release if possible, especially since the current algorithm has so many wrong answers.

**Comment by Upabjojr** [comment]
There are two cases from the cookbook which are currently giving a wrong result. I haven't investigated why yet. I'll try this weekend.

File: sympy/matrices/expressions/tests/test_derivatives.py (hunk 5 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: It looks correct. I have replaced the matrix symbols with entry matrices using SymEngine and then waited for hours for the computations to perform. I will share the script later.

**Comment by asmeurer** [comment]
Interesting. Was your old code also giving the same wrong result as the cookbook?

**Comment by Upabjojr** [comment]
> Interesting. Was your old code also giving the same wrong result as the cookbook?

The old code was returning 0 in example 125 (which is very wrong) and raising an exception in example 127.

This code is definitely way more robust than the previous one.

File: sympy/matrices/expressions/tests/test_derivatives.py [hunk 6]
```diff
@@ -189,3 +200,43 @@ def test_matrix_derivatives_of_traces():
...
+    # Other
+
+    # Cookbook example 124:
+    expr = Trace(A*X**(-1)*B)
+    assert expr.diff(X) == -Inverse(X).T*A.T*B.T*Inverse(X).T
+
+    # Cookbook example 125:
+    expr = Trace(Inverse(X.T*C*X)*A)
+    # Warning: result in the cookbook appears to be wrong:
+    assert expr.diff(X) == - X.inv().T*A.T*X.inv()*C.inv().T*X.inv().T - X.inv().T*A*X.inv()*C.inv()*X.inv().T
```
[the remark this answers is not in the document]
Comment (jksuom) [reply sympy/matrices/expressions/tests/test_derivatives.py]: Cookbook assumes that C is symmetric: `C == C.T`.

[the remark this answers is not in the document]
Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: Thanks for the point, I missed that detail.

**Comment by Upabjojr** [comment]
I would go for a merge as soon as tests pass. @asmeurer OK?

**Comment by Upabjojr** [comment]
The tests 125 and 127 are equivalent with the results in the cookbook if you assume that `B` and `C` are symmetric, which solves the mystery.

**Comment by asmeurer** [comment]
This looks good. I hope the lines idea can be easily applied to all differentiable matrix functions.

**Comment by benelot** [comment]
Hello! I am currently trying to turn the derivations of the computational models of my research group into jupyter notebooks. Sympy seems to be the de-facto standard to do symbolic manipulation. While I am implementing our formulas (Matrix-Vector equations, derived by vectors). I got some unusual results (that are different from those on matrixcalculus.org). I was not able to find out if this here is the same algorithm as matrixcalculus.org, so I could not tell. As I am playing around with it a lot, I could provide some failing test cases if that is appropriate. Is there a list of derivative type that should be working? Should I point out the weird results in issues? Should I point out cases that differ from matrixcalculus.org?

Just an aside question: Derivative(_, _).doit() should always give the same result as diff(_, _) or _.diff(_), right? (this was not true in some cases of sympy 1.7.1, before I realized there is an upgrade).

**Comment by oscarbenjamin** [comment]
Can you list the unusual results in a new issue so that they can be checked?

**Comment by asmeurer** [comment]
You may also want to CC @Upabjojr on any issues with matrix derivatives. He has been the primary one working on it.

**Comment by Upabjojr** [comment]
> While I am implementing our formulas (Matrix-Vector equations, derived by vectors). I got some unusual results (that are different from those on matrixcalculus.org). 

Possible, matrixcalculus.org expressions support broadcasting, we don't (i.e. if a matrix is a vector, the trivial dimension can be multiplied with any matrix dimension, in SymPy it can only be multiplied with another trivial dimension). Actually, it's possible that some of matrixcalculus.org results are not even valid in SymPy.

> I was not able to find out if this here is the same algorithm as matrixcalculus.org, so I could not tell.

No, it's completely different.

> As I am playing around with it a lot, I could provide some failing test cases if that is appropriate. Is there a list of derivative type that should be working? Should I point out the weird results in issues? Should I point out cases that differ from matrixcalculus.org?

There is a better algorithm for array derivatives in `sympy.tensor.array.expressions`, hopefully it will replace the curret matrix derivation algorithm in the future.

**Comment by benelot** [comment]
> There is a better algorithm for array derivatives in `sympy.tensor.array.expressions`, hopefully it will replace the curret matrix derivation algorithm in the future.

Thanks for the answer. Is there a practical way to use the `sympy.tensor.array.expressions` algorithm for matrix-vector derivation? I am still looking for a way to derive by vectors and undefined vector functions d/du (1/2 (u - W r(u))^T * (u - W r(u))) =  u - W r(u) - r'(u) \odot (W^T (u - W r(u))) (the rather trivial case), but I struggle to get it working properly.

**Comment by Upabjojr** [comment]
You may want to try `ArrayElementwiseApplyFunc` combined with `array_derive`.

**Comment by Upabjojr** [comment]
`convert_matrix_to_array` and `convert_array_to_matrix` may also help.

**Pull request merged**